subsys/usb/usb_device.c

```diff
@@ -1004,3 +1040,247 @@ int usb_ep_read_continue(u8_t ep)
 {
 	return usb_dc_ep_read_continue(ep);
 }
+
+/* Transfer management */
+static struct usb_transfer_data *usb_ep_get_transfer(u8_t ep)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(usb_dev.transfer); i++) {
+		if (usb_dev.transfer[i].ep == ep) {
+			return &usb_dev.transfer[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void usb_transfer_work(struct k_work *item)
+{
+	struct usb_transfer_data *trans;
+	int ret = 0, bytes;
+	u8_t ep;
+
+	trans = CONTAINER_OF(item, struct usb_transfer_data, work);
+	ep = trans->ep;
+
+	if (trans->status != -EBUSY) {
+		/* transfer cancelled or already completed */
+		goto done;
+	}
+
+	if (trans->flags & USB_TRANS_WRITE) {
```

Comment:
Sorry if I did not get the code right, but should we have here while(true){} loop? we can break loop (instead of goto) on !trans->bsize

Reply:
'''
diff --git a/subsys/usb/usb_device.c b/subsys/usb/usb_device.c
index ce94b9ddf4..a5639e0d4a 100644
--- a/subsys/usb/usb_device.c
+++ b/subsys/usb/usb_device.c
@@ -1073,6 +1073,8 @@ static void usb_transfer_work(struct k_work *item)
                goto done;
        }
 
+       while (true) {
+
        if (trans->flags & USB_TRANS_WRITE) {
                if (!trans->bsize) {
                        if (!(trans->flags & USB_TRANS_NO_ZLP)) {
@@ -1116,6 +1118,8 @@ static void usb_transfer_work(struct k_work *item)
                usb_dc_ep_read_continue(ep);
        }
 
+       }
+
 done:
        if (trans->status != -EBUSY && trans->cb) { /* Transfer complete */
                usb_transfer_callback cb = trans->cb;
'''

Reply:
Why introducing this while ? the transfer work is 'normally' called each time we have data to read or to write, so that we are driven by events without active polling.

Reply:
If you call usb_transfer_sync() with WRITE operation callback never gets called without while loop.

Reply:
This is what I used in the EEM rework (included in this PR), it was working. Do you have a specific sample I could try? On device write operation complete, device driver calls the endpoint callback from which we call this work, then we can continue writing the next data chunk and so on.

Reply:
Let's say we have one small data chunk (i.e. Bluetooth), you have a check for `!trans->bsize` before actual write, so that `trans->status == -EBUSY` and callback does not get called. You can change hci_usb sample with:

```
                 switch (bt_buf_get_type(buf)) {
                case BT_BUF_EVT:
-                       try_write(BTUSB_ENDP_INT, buf);
+                       usb_transfer_sync(BTUSB_ENDP_INT, buf->data, buf->len,
+                                         USB_TRANS_WRITE);
                        break;
                case BT_BUF_ACL_IN:
-                       try_write(BTUSB_ENDP_BULK_IN, buf);
+                       usb_transfer_sync(BTUSB_ENDP_BULK_IN, buf->data,
+                                         buf->len, USB_TRANS_WRITE);
                        break;
                default:
                        SYS_LOG_ERR("Unknown type %u", bt_buf_get_type(buf));
                        break;
                }
```

Reply:
Oh I see, Actually when you want to use transfer method, you have to configure your endpoint accordingly by setting the transfer callback, this is something which has been discussed with @jfischer-phytec-iot previously, but maybe the documentation in usb_device.h should be clarified:
```

--- a/samples/bluetooth/hci_usb/src/main.c
+++ b/samples/bluetooth/hci_usb/src/main.c
@@ -480,7 +480,7 @@ static void btusb_iso_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
 /* Describe EndPoints configuration */
 static struct usb_ep_cfg_data btusb_ep[] = {
        {
-               .ep_cb  = btusb_int_in,
+               .ep_cb  = usb_transfer_ep_callback,
                .ep_addr = BTUSB_ENDP_INT
        },
        {
@@ -488,7 +488,7 @@ static struct usb_ep_cfg_data btusb_ep[] = {
                .ep_addr = BTUSB_ENDP_BULK_OUT
        },
        {
-               .ep_cb = btusb_bulk_in,
+               .ep_cb = usb_transfer_ep_callback,
                .ep_addr = BTUSB_ENDP_BULK_IN
        },
        {
@@ -689,10 +689,12 @@ void main(void)
 
                switch (bt_buf_get_type(buf)) {
                case BT_BUF_EVT:
-                       try_write(BTUSB_ENDP_INT, buf);
+                       usb_transfer_sync(BTUSB_ENDP_INT, buf->data, buf->len,
+                                         USB_TRANS_WRITE);
                        break;
                case BT_BUF_ACL_IN:
-                       try_write(BTUSB_ENDP_BULK_IN, buf);
+                       usb_transfer_sync(BTUSB_ENDP_BULK_IN, buf->data,
+                                         buf->len, USB_TRANS_WRITE);
                        break;
                default:
                        SYS_LOG_ERR("Unknown type %u", bt_buf_get_type(buf));
```